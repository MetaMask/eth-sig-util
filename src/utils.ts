import {
  addHexPrefix,
  bufferToHex,
  bufferToInt,
  ecrecover,
  fromRpcSig,
  fromSigned,
  toBuffer,
  ToBufferInputTypes,
  toUnsigned,
} from '@ethereumjs/util';
import { intToHex, isHexString, stripHexPrefix } from 'ethjs-util';

/**
 * Pads the front of the given hex string with zeroes until it reaches the
 * target length. If the input string is already longer than or equal to the
 * target length, it is returned unmodified.
 *
 * If the input string is "0x"-prefixed or not a hex string, an error will be
 * thrown.
 *
 * @param hexString - The hexadecimal string to pad with zeroes.
 * @param targetLength - The target length of the hexadecimal string.
 * @returns The input string front-padded with zeroes, or the original string
 * if it was already greater than or equal to to the target length.
 */
export function padWithZeroes(hexString: string, targetLength: number): string {
  if (hexString !== '' && !/^[a-f0-9]+$/iu.test(hexString)) {
    throw new Error(
      `Expected an unprefixed hex string. Received: ${hexString}`,
    );
  }

  if (targetLength < 0) {
    throw new Error(
      `Expected a non-negative integer target length. Received: ${targetLength}`,
    );
  }

  return String.prototype.padStart.call(hexString, targetLength, '0');
}

/**
 * Returns `true` if the given value is nullish.
 *
 * @param value - The value being checked.
 * @returns Whether the value is nullish.
 */
export function isNullish(value) {
  return value === null || value === undefined;
}

/**
 * Convert a value to a Buffer. This function should be equivalent to the `toBuffer` function in
 * `ethereumjs-util@5.2.1`.
 *
 * @param value - The value to convert to a Buffer.
 * @returns The given value as a Buffer.
 */
export function legacyToBuffer(value: ToBufferInputTypes) {
  return typeof value === 'string' && !isHexString(value)
    ? Buffer.from(value)
    : toBuffer(value);
}

/**
 * Concatenate an extended ECDSA signature into a single '0x'-prefixed hex string.
 *
 * @param v - The 'v' portion of the signature.
 * @param r - The 'r' portion of the signature.
 * @param s - The 's' portion of the signature.
 * @returns The concatenated ECDSA signature as a '0x'-prefixed string.
 */
export function concatSig(v: Buffer, r: Buffer, s: Buffer): string {
  const rSig = fromSigned(r);
  const sSig = fromSigned(s);
  const vSig = bufferToInt(v);
  const rStr = padWithZeroes(toUnsigned(rSig).toString('hex'), 64);
  const sStr = padWithZeroes(toUnsigned(sSig).toString('hex'), 64);
  const vStr = stripHexPrefix(intToHex(vSig));
  return addHexPrefix(rStr.concat(sStr, vStr));
}

/**
 * Recover the public key from the given signature and message hash.
 *
 * @param messageHash - The hash of the signed message.
 * @param signature - The signature.
 * @returns The public key of the signer.
 */
export function recoverPublicKey(
  messageHash: Buffer,
  signature: string,
): Buffer {
  const sigParams = fromRpcSig(signature);
  return ecrecover(messageHash, sigParams.v, sigParams.r, sigParams.s);
}

/**
 * Normalize the input to a lower-cased '0x'-prefixed hex string.
 *
 * @param input - The value to normalize.
 * @returns The normalized value.
 */
export function normalize(input: number | string): string {
  if (!input) {
    return undefined;
  }

  if (typeof input === 'number') {
    if (input < 0) {
      return '0x';
    }
    const buffer = toBuffer(input);
    input = bufferToHex(buffer);
  }

  if (typeof input !== 'string') {
    let msg = 'eth-sig-util.normalize() requires hex string or integer input.';
    msg += ` received ${typeof input}: ${input}`;
    throw new Error(msg);
  }

  return addHexPrefix(input.toLowerCase());
}

/**
 * Node's Buffer.from() method does not seem to buffer numbers correctly out of the box.
 * This helper method formats the number correct for Buffer.from to return correct buffer.
 *
 * @param num - The number to convert to buffer.
 * @returns The number in buffer form.
 */
export function numberToBuffer(num: number) {
  const hexVal = num.toString(16);
  const prepend = hexVal.length % 2 ? '0' : '';
  return Buffer.from(prepend + hexVal, 'hex');
}
