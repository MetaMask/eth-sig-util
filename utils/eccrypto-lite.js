'use strict'

const EC = require('elliptic').ec

const ec = new EC('secp256k1')
const cryptoObj = global.crypto || global.msCrypto || {}
const subtle = cryptoObj.subtle || cryptoObj.webkitSubtle

function assert (condition, message) {
  if (!condition) {
    throw new Error(message || 'Assertion failed')
  }
}

function randomBytes (size) {
  const arr = new Uint8Array(size)
  global.crypto.getRandomValues(arr)
  return Buffer.from(arr)
}

function sha512 (msg) {
  return subtle.digest({ name: 'SHA-512' }, msg).then(function (hash) {
    return Buffer.from(new Uint8Array(hash))
  })
}

function getAes (op) {
  return function (iv, key, data) {
    const importAlgorithm = { name: 'AES-CBC' }
    const keyp = subtle.importKey('raw', key, importAlgorithm, false, [op])
    return keyp.then(function (cryptoKey) {
      const encAlgorithm = { name: 'AES-CBC', iv }
      return subtle[op](encAlgorithm, cryptoKey, data)
    }).then(function (result) {
      return Buffer.from(new Uint8Array(result))
    })
  }
}

const aesCbcEncrypt = getAes('encrypt')
const aesCbcDecrypt = getAes('decrypt')

function hmacSha256Sign (key, msg) {
  const algorithm = { name: 'HMAC', hash: { name: 'SHA-256' } }
  const keyp = subtle.importKey('raw', key, algorithm, false, ['sign'])
  return keyp.then(function (cryptoKey) {
    return subtle.sign(algorithm, cryptoKey, msg)
  }).then(function (sig) {
    return Buffer.from(new Uint8Array(sig))
  })
}

function hmacSha256Verify (key, msg, sig) {
  const algorithm = { name: 'HMAC', hash: { name: 'SHA-256' } }
  const keyp = subtle.importKey('raw', key, algorithm, false, ['verify'])
  return keyp.then(function (cryptoKey) {
    return subtle.verify(algorithm, cryptoKey, sig, msg)
  })
}

const getPublic = function (privateKey) {
  assert(privateKey.length === 32, 'Bad private key')
  return Buffer.from(ec.keyFromPrivate(privateKey).getPublic('arr'))
}
exports.getPublic = getPublic

const derive = function (privateKeyA, publicKeyB) {
  return new Promise(function (resolve) {
    assert(Buffer.isBuffer(privateKeyA), 'Bad input')
    assert(Buffer.isBuffer(publicKeyB), 'Bad input')
    assert(privateKeyA.length === 32, 'Bad private key')
    assert(publicKeyB.length === 65, 'Bad public key')
    assert(publicKeyB[0] === 4, 'Bad public key')
    const keyA = ec.keyFromPrivate(privateKeyA)
    const keyB = ec.keyFromPublic(publicKeyB)
    const Px = keyA.derive(keyB.getPublic()) // BN instance
    resolve(Buffer.from(Px.toArray()))
  })
}
exports.derive = derive

const encrypt = function (publicKeyTo, msg, opts) {
  assert(subtle, 'WebCryptoAPI is not available')
  opts = opts || {}
  // Tmp variables to save context from flat promises;
  let iv, ephemPublicKey, ciphertext, macKey
  return new Promise(function (resolve) {
    const ephemPrivateKey = opts.ephemPrivateKey || randomBytes(32)
    ephemPublicKey = getPublic(ephemPrivateKey)
    resolve(derive(ephemPrivateKey, publicKeyTo))
  }).then(function (Px) {
    return sha512(Px)
  }).then(function (hash) {
    iv = opts.iv || randomBytes(16)
    const encryptionKey = hash.slice(0, 32)
    macKey = hash.slice(32)
    return aesCbcEncrypt(iv, encryptionKey, msg)
  }).then(function (data) {
    ciphertext = data
    const dataToMac = Buffer.concat([iv, ephemPublicKey, ciphertext])
    return hmacSha256Sign(macKey, dataToMac)
  }).then(function (mac) {
    return {
      iv,
      ephemPublicKey,
      ciphertext,
      mac,
    }
  })
}
const decrypt = function (privateKey, opts) {
  assert(subtle, 'WebCryptoAPI is not available')
  // Tmp variable to save context from flat promises;
  let encryptionKey
  return derive(privateKey, opts.ephemPublicKey).then(function (Px) {
    return sha512(Px)
  }).then(function (hash) {
    encryptionKey = hash.slice(0, 32)
    const macKey = hash.slice(32)
    const dataToMac = Buffer.concat([
      opts.iv,
      opts.ephemPublicKey,
      opts.ciphertext,
    ])
    return hmacSha256Verify(macKey, dataToMac, opts.mac)
  }).then(function (macGood) {
    assert(macGood, 'Bad MAC')
    return aesCbcDecrypt(opts.iv, encryptionKey, opts.ciphertext)
  }).then(function (msg) {
    return Buffer.from(new Uint8Array(msg))
  })
}

module.exports = {
  async decryptWithPrivateKey (privateKey, encrypted) {
    console.log(privateKey)
    const twoStripped = privateKey.replace(/^.{2}/gu, '')
    console.log(twoStripped)
    const encryptedBuffer = {
      iv: Buffer.from(encrypted.iv, 'hex'),
      ephemPublicKey: Buffer.from(encrypted.ephemPublicKey, 'hex'),
      ciphertext: Buffer.from(encrypted.ciphertext, 'hex'),
      mac: Buffer.from(encrypted.mac, 'hex'),
    }

    const decryptedBuffer = await decrypt(
      Buffer.from(twoStripped, 'hex'),
      encryptedBuffer,
    )
    return decryptedBuffer.toString()


  },
  async encryptWithPublicKey (receiverPublicKey, payload) {
    const pubString = `04${receiverPublicKey}`
    const encryptedBuffers = await encrypt(
      Buffer.from(pubString, 'hex'),
      Buffer.from(payload),
    )
    const encrypted = {
      iv: encryptedBuffers.iv.toString('hex'),
      ephemPublicKey: encryptedBuffers.ephemPublicKey.toString('hex'),
      ciphertext: encryptedBuffers.ciphertext.toString('hex'),
      mac: encryptedBuffers.mac.toString('hex'),
    }
    return encrypted
  },
}
