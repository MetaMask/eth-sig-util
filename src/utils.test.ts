import { concatSig, padWithZeroes, normalize } from './utils';

describe('padWithZeroes', function () {
  it('pads a string shorter than the target length with zeroes', function () {
    const input = 'abc';
    expect(padWithZeroes(input, 5)).toStrictEqual(`00${input}`);
  });

  it('pads an empty string', function () {
    const input = '';
    expect(padWithZeroes(input, 4)).toStrictEqual(`0000`);
  });

  it('returns a string equal to the target length without modifying it', function () {
    const input = 'abc';
    expect(padWithZeroes(input, 3)).toStrictEqual(input);
  });

  it('returns a string longer than the target length without modifying it', function () {
    const input = 'abcd';
    expect(padWithZeroes(input, 3)).toStrictEqual(input);
  });

  it('throws an error if passed an invalid hex string', function () {
    const inputs = ['0xabc', 'xyz', '-'];
    for (const input of inputs) {
      expect(() => padWithZeroes(input, 3)).toThrow(
        new Error(`Expected an unprefixed hex string. Received: ${input}`),
      );
    }
  });

  it('throws an error if passed a negative number', function () {
    expect(() => padWithZeroes('abc', -1)).toThrow(
      new Error('Expected a non-negative integer target length. Received: -1'),
    );
  });
});

describe('concatSig', function () {
  it('should concatenate an extended ECDSA signature', function () {
    expect(
      concatSig(
        Buffer.from('1', 'hex'),
        Buffer.from('1', 'hex'),
        Buffer.from('1', 'hex'),
      ),
    ).toMatchInlineSnapshot(
      `"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`,
    );
  });

  it('should concatenate an all-zero extended ECDSA signature', function () {
    expect(
      concatSig(
        Buffer.from('0', 'hex'),
        Buffer.from('0', 'hex'),
        Buffer.from('0', 'hex'),
      ),
    ).toMatchInlineSnapshot(
      `"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`,
    );
  });

  it('should return a hex-prefixed string', function () {
    const signature = concatSig(
      Buffer.from('1', 'hex'),
      Buffer.from('1', 'hex'),
      Buffer.from('1', 'hex'),
    );

    expect(typeof signature).toBe('string');
    expect(signature.slice(0, 2)).toBe('0x');
  });

  it('should encode an impossibly large extended ECDSA signature', function () {
    const largeNumber = Number.MAX_SAFE_INTEGER.toString(16);
    expect(
      concatSig(
        Buffer.from(largeNumber, 'hex'),
        Buffer.from(largeNumber, 'hex'),
        Buffer.from(largeNumber, 'hex'),
      ),
    ).toMatchInlineSnapshot(
      `"0x000000000000000000000000000000000000000000000000001fffffffffffff000000000000000000000000000000000000000000000000001fffffffffffff1fffffffffffff"`,
    );
  });

  it('should throw if a portion of the signature is larger than the maximum safe integer', function () {
    const largeNumber = '20000000000000'; // This is Number.MAX_SAFE_INTEGER + 1, in hex
    expect(() =>
      concatSig(
        Buffer.from(largeNumber, 'hex'),
        Buffer.from(largeNumber, 'hex'),
        Buffer.from(largeNumber, 'hex'),
      ),
    ).toThrow('Number exceeds 53 bits');
  });
});

describe('normalize', function () {
  it('should normalize an address to lower case', function () {
    const initial = '0xA06599BD35921CfB5B71B4BE3869740385b0B306';
    const result = normalize(initial);
    expect(result).toBe(initial.toLowerCase());
  });

  it('should normalize address without a 0x prefix', function () {
    const initial = 'A06599BD35921CfB5B71B4BE3869740385b0B306';
    const result = normalize(initial);
    expect(result).toBe(`0x${initial.toLowerCase()}`);
  });

  it('should normalize an integer to a byte-pair hex string', function () {
    const initial = 1;
    const result = normalize(initial);
    expect(result).toBe('0x01');
  });

  // TODO: Add validation to disallow negative integers.
  it('should normalize a negative integer to 0x', function () {
    const initial = -1;
    const result = normalize(initial);
    expect(result).toBe('0x');
  });

  // TODO: Add validation to disallow null.
  it('should return undefined if given null', function () {
    const initial = null;
    expect(normalize(initial as any)).toBeUndefined();
  });

  // TODO: Add validation to disallow undefined.
  it('should return undefined if given undefined', function () {
    const initial = undefined;
    expect(normalize(initial as any)).toBeUndefined();
  });

  it('should throw if given an object', function () {
    const initial = {};
    expect(() => normalize(initial as any)).toThrow(
      'eth-sig-util.normalize() requires hex string or integer input. received object:',
    );
  });

  it('should throw if given a boolean', function () {
    const initial = true;
    expect(() => normalize(initial as any)).toThrow(
      'eth-sig-util.normalize() requires hex string or integer input. received boolean: true',
    );
  });

  it('should throw if given a bigint', function () {
    const initial = BigInt(Number.MAX_SAFE_INTEGER);
    expect(() => normalize(initial as any)).toThrow(
      'eth-sig-util.normalize() requires hex string or integer input. received bigint: 9007199254740991',
    );
  });

  it('should throw if given a symbol', function () {
    const initial = Symbol('test');
    expect(() => normalize(initial as any)).toThrow(
      'Cannot convert a Symbol value to a string',
    );
  });
});
