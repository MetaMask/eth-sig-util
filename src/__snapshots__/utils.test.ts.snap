// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concatSig should concatenate an all-zero extended ECDSA signature 1`] = `"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`concatSig should concatenate an extended ECDSA signature 1`] = `"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`concatSig should encode an impossibly large extended ECDSA signature 1`] = `"0x000000000000000000000000000000000000000000000000001fffffffffffff000000000000000000000000000000000000000000000000001fffffffffffff1fffffffffffff"`;
